### DISCUSSION TITLE
How do I detect if a pdf contains text that has been rotated?

### DISCUSSION THREAD

**Discussion by cabal-chan** [pymupdf/PyMuPDF#916] [Q&A] [answered]
I've been looking through the docs and I can't seem to find anything on here that allows for this? This is causing issues when adding annotations to the document. How I see addressing this would be to detect if this needs to be rotated, rotate it to that amount, add annotations, and rotate it back. What's the best way to address this issue?

See page 261. 
[ncaa.pdf](https://github.com/pymupdf/PyMuPDF/files/6019407/ncaa.pdf)

**Comment by JorjMcKie** [comment, marked as answer]
Before I take a look at the PDF, some general comments:
* Text extraction (via "dict", etc.) only returns the **_bbox_** ( = rectangle) of the text not the quadrilateral. For what that means look at this picture from the documentation:
![grafik](https://user-images.githubusercontent.com/8290722/108693897-c449ed00-74d4-11eb-832b-856cbc83b529.png)
* I am returning `(x0,y0, x1, y1)` corresponding to the blue rectangle, **_not_** `(ul, ur, ll, lr)` corresponding to the corners of the quad. I do consider introducing an option to request a different output in some future version.
* There is an information contained in the "dict" output that at least lets you check whether there is a rotation: this is the `dir` key of the line sub-dictionary: 
![grafik](https://user-images.githubusercontent.com/8290722/108694842-f740b080-74d5-11eb-9d4d-75ad40b66254.png)
* In cases of 90 degree rotations (= exactly one component of `dir` is zero), the quad **_indeed is_** sufficiently specified by the coordinates x0, y0, x1, y1 and - with some coordinate juggling - you can correctly mark the text with an annotation - see this description:

Text extraction with the “dict” option of `Page.get_text()` may also return text with a non-zero angle to the x-axis. This is reflected by the value of the "dir" key of the line dictionary: it is the tuple (cosine, sine) of that angle.

Currently, all bboxes returned by the method’s are rectangles only – no quads. So we can mark the span text (correctly) only, if the angle is 0, 90, 180 or 270 degrees.

In this case we can convert the span bbox into the right quad by choosing the right sequence of its corners:

```python
r = fitz.Rect(span["bbox"])

if line["dir"] == (1, 0):  # rotation 0
    q = fitz.Quad(r)

elif line["dir"] == (0, -1):  # rotation 90
    q = fitz.Quad(r.bl, r.tl, r.br, r.tr)

elif line["dir"] == (-1, 0):  # rotation 180
    q = fitz.Quad(r.br, r.bl, r.tr, r.tl)

elif line["dir"] == (0, 1):  # rotation 270
    q = fitz.Quad(r.tr, r.br, r.tl, r.bl)

else:
    q = fitz.Quad(r)
    print("warning: unsupported text flow")

annot = page.addHighlightAnnot(q)
```

**Reply by cabal-chan** [reply]
Thanks @JorjMcKie! I did stumble upon this after some more research but forgot to update the issue. Will try this approach next and see how it goes!
